components/brave_wallet/browser/sns_resolver_task.cc

```diff
@@ -0,0 +1,665 @@
+SnsResolverTaskError::SnsResolverTaskError(SnsResolverTaskError&&) = default;
+SnsResolverTaskError& SnsResolverTaskError::operator=(
+    const SnsResolverTaskError&) = default;
+SnsResolverTaskError& SnsResolverTaskError::operator=(SnsResolverTaskError&&) =
+    default;
+SnsResolverTaskError::~SnsResolverTaskError() = default;
+
+class ScopedWorkOnSnsTask {
+ public:
+  explicit ScopedWorkOnSnsTask(SnsResolverTask* task) : task_(task) {}
+  ~ScopedWorkOnSnsTask() { task_->WorkOnTask(); }
+
+ private:
+  raw_ptr<SnsResolverTask> task_ = nullptr;
+};
+
+SnsResolverTask::SnsResolverTask(DoneCallback done_callback,
+                                 APIRequestHelper* api_request_helper,
+                                 const std::string& domain,
+                                 const GURL& network_url)
+    : done_callback_(std::move(done_callback)),
+      api_request_helper_(api_request_helper),
+      domain_(domain),
+      network_url_(network_url) {
+  DCHECK(api_request_helper_);
+}
+
+SnsResolverTask::~SnsResolverTask() = default;
+
+// static
+base::RepeatingCallback<void(SnsResolverTask* task)>&
+SnsResolverTask::GetWorkOnTaskForTesting() {
+  static base::NoDestructor<
+      base::RepeatingCallback<void(SnsResolverTask * task)>>
+      callback;
+  return *callback.get();
+}
+
+void SnsResolverTask::SetResultForTesting(
+    absl::optional<SnsResolverTaskResult> task_result,
+    absl::optional<SnsResolverTaskError> task_error) {
+  task_result_ = std::move(task_result);
+  task_error_ = std::move(task_error);
+}
+
+void SnsResolverTask::ScheduleWorkOnTask() {
+  base::SequencedTaskRunnerHandle::Get()->PostTask(
+      FROM_HERE, base::BindOnce(&SnsResolverTask::WorkOnTask,
+                                weak_ptr_factory_.GetWeakPtr()));
+}
+
+void SnsResolverTask::WorkOnTask() {
+  if (!GetWorkOnTaskForTesting().is_null()) {
+    GetWorkOnTaskForTesting().Run(this);
+  }
+
+  if (task_result_) {
+    std::move(done_callback_).Run(this, std::move(task_result_), absl::nullopt);
+    // `this` is not valid here
+    return;
+  }
+  if (task_error_) {
+    std::move(done_callback_).Run(this, absl::nullopt, std::move(task_error_));
+    // `this` is not valid here.
+    return;
+  }
+
+  if (!domain_address_) {
+    domain_address_ = GetDomainKey(domain_, false);
+    if (!domain_address_) {
+      SetError(MakeInvalidParamsError());
+      ScheduleWorkOnTask();
+      return;
+    }
+  }
+
+  // Check if there is an nft token associated with domain.
+  if (!nft_owner_check_done_) {
+    if (!nft_mint_address_) {
+      nft_mint_address_ = GetMintAddress(*domain_address_);
+      if (!nft_mint_address_) {
+        NftOwnerDone(absl::nullopt);
+        ScheduleWorkOnTask();
+        return;
+      }
+    }
+
+    // Check if domain is minted and its supply is 1.
+    if (!nft_mint_supply_check_done_) {
+      FetchNftSplMint();
+      return;
+    }
+
+    // Find token account for that mint and extract owner from its data.
```

Comment:
// Find token account for that mint and extract owner from its data to be the resolution result.

Reply:
https://github.com/brave/brave-core/pull/15617/commits/1e34bd9363c2e1aa4593c00f3409b8c119ccab29